File: FieldGuide/9_Python/grade.cpp
#include "splashkit.h"
#include "utilities.h"
#include "splashkit-arrays.h"

class grade_analyser
{

    // define a list of a grade(empty)
    dynamic_array<int> grades{};

public:
    // add one grade data
    void add_grade(int g)
    {
        add(grades, g);
    }

    // add a list of grades
    // overload add_grade()
    // need array
    void add_grade(dynamic_array<int> g)
    {
        for (int i = 0; i < length(g); i++)
        {
            add_grade(g[i]);
        }
    }

    // another version of overload
    // no need array
    void add_grade()
    {
        int count = read_integer("How many grades? ");
        for (int i = 0; i < count; i++)
        {
            int g = read_integer("Enter grade " + to_string(i + 1) + ": ");
            add_grade(g);
        }
    }

    // average
    double average()
    {
        double total{};
        for (int i = 0; i < length(grades); i++)
        {
            total += grades[i];
        }
        return total / length(grades);
    }

    // pass rate
    double pass_rate()
    {
        // the num of student grades >= 50
        int num_pass{};
        for (int i = 0; i < length(grades); i++)
        {
            if (grades[i] >= 50)
            {
                num_pass++;
            }
        }
        return (double)num_pass / length(grades) * 100;
    }

    // return the number of student is hd
    int num_hd()
    {
        int n_hd{};
        for (int i = 0; i < length(grades); i++)
        {
            if (grades[i] >= 80)
            {
                n_hd++;
            }
        }
        return n_hd;
    }
};

int main()
{
    grade_analyser analyser;

    // Let the user enter a list of grades
    analyser.add_grade();

    // Output the results
    write_line("Average: " + to_string(analyser.average()));
    write_line("Pass rate: " + to_string(analyser.pass_rate() * 100) + "%");
    write_line("High distinctions: " + to_string(analyser.num_hd()));

    return 0;
}

File: FieldGuide/P8_Python/grade.py
class grade_analyser:
    def __init__(self):
        self.grades = []

    #add one grade
    def add_grade(self, g):
        self.grades.append(g)

    #add many grades
    def add_grades(self):
        count = int(input("How many grades? "))
        for i in range(count):
            g = int(input(f"Grade {i + 1} : "))
            self.grades.append(g)

    #add grades by list
    def add_grades_list(self, g):
        self.grades += g

    #average
    def average(self):
        total = 0
        for g in self.grades:
            total += g
        return total / len(self.grades)

    #pass_rate
    def pass_rate(self):
        num_pass = 0
        for g in self.grades:
            if g >= 50:
                num_pass += 1
        return num_pass / len(self.grades) * 100

    #num_hd
    def num_hd(self):
        num_hd = 0
        for g in self.grades:
            if g >= 80:
                num_hd += 1
        return num_hd

if __name__ == "__main__":
    analyser = grade_analyser()
    analyser.add_grades()

    print(f"Average: {analyser.average()}")
    print(f"Pass rate: {analyser.pass_rate()}%")
    print(f"High distinctions: {analyser.num_hd()}")